agents/hwp_rotation/src/pid_controller.py

```diff
@@ -0,0 +1,188 @@
+        self.return_list = []
+        self.PID_INFO = [pid_ip, pid_port]
+        self.hex_freq = '00000'
+        self.cur_freq = 0
+        self.stopping = True
+        self.stop_params = [0.2, 0, 0]
+        self.tune_params = [0.2, 63, 0]
+
+
+########################################################################################################################
+# Subprocesses
+########################################################################################################################
+
+    # Converts the user input inrto a format the PID controller can read
+    def convert_to_hex(self, value, decimal):
+        temp_value = hex(int(10**decimal*float(value)))
+        return ('0000' + str(temp_value)[2:].upper())[-4:]
+
+    # Opens the connection with the PID controller and makes sure that nothing else is using the connection
+    def open_line(self):
+        while True:
+            try:
+                self.lock_file = open(os.path.join(this_dir, '.pid_port_busy'))
+                fcntl.flock(self.lock_file, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                break
+            except BlockingIOError:
+                time.sleep(2)
+
+    # Closes the connection with the PID controller
+    def close_line(self):
+        fcntl.flock(self.lock_file, fcntl.LOCK_UN)
+        self.lock_file.close()
+
+########################################################################################################################
+# Main Processes
+########################################################################################################################
+
+    # Sets the direction if the CHWP; 0 for forward and 1 for backwards
+    def set_direction(self, direction):
+        self.open_line()   
+        subprocess.call([os.path.join(self.script_dir, 'tune_direction'), self.PID_INFO[0], self.PID_INFO[1],
+                         direction], stderr = subprocess.DEVNULL)
+        if direction == '0':
+            print('Forward')
+            self.stopping = False
+        elif direction == '1':
+            print('Reverse')
+            self.stopping = True
+    
+        self.return_messages()
+        self.close_line()
+
+    # Declare to memory what the CHWP frequency should be (does not actually change the frequency)
+    def declare_freq(self, freq):
+        if float(freq) <= 3.5:
+            self.hex_freq = '0' + self.convert_to_hex(freq, 3)
+            if self.verb:
+                print('Frequency Setpoint = ' + str(freq) + ' Hz')
+        else:
+            if self.verb:
+                print('Invalid Frequency')
+
+    # Method which sets the setpoint to 0 Hz and stops the CHWP
+    def tune_stop(self):
+        self.open_line()
+        if self.verb:
+            print('Starting Stop')
+        subprocess.call([os.path.join(self.script_dir, 'tune_stop'), self.PID_INFO[0], self.PID_INFO[1]], 
+                         stderr = subprocess.DEVNULL)
```

Comment:
Rather than subprocess all the calls to bash scripts that run telnet I think you should consider using [telnetlib](https://docs.python.org/3/library/telnetlib.html) directly in Python.

In either case, some error handling would be good. What happens if the `tune_stop` script can't communicate with the device?

Reply:
telnet doesn't seem to be a requirement here, so I've replaced the subprocess calls to bash scripts calling telnet with commands send via a socket connection in #281.